File: frontend/app/javafx/src/main/java/io/mateu/javafx/ui/renderer/PageRenderer.java
package io.mateu.javafx.ui.renderer;

import com.fasterxml.jackson.databind.JsonNode;
import io.mateu.javafx.ui.AppContext;
import javafx.geometry.Insets;
import javafx.scene.Node;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.ScrollPane;
import javafx.scene.layout.*;

public class PageRenderer {

    private final AppContext ctx;

    public PageRenderer(AppContext ctx) {
        this.ctx = ctx;
    }

    public Node render(JsonNode component, JsonNode metadata, JsonNode state, JsonNode data) {
        BorderPane root = new BorderPane();
        root.getStyleClass().add("content-area");

        // Header
        String title = metadata.path("title").asText("");
        String subtitle = metadata.path("subtitle").asText("");
        String pageTitle = metadata.path("pageTitle").asText("");
        String headerText = !title.isBlank() ? title : pageTitle;

        JsonNode toolbar = metadata.path("toolbar");
        boolean hasToolbar = toolbar.isArray() && !toolbar.isEmpty();

        if (!headerText.isBlank() || !subtitle.isBlank() || hasToolbar) {
            VBox header = new VBox(4);
            header.setPadding(new Insets(0, 0, 16, 0));
            if (!headerText.isBlank()) {
                Label titleLabel = new Label(headerText);
                titleLabel.getStyleClass().add("form-title");
                header.getChildren().add(titleLabel);
            }
            if (!subtitle.isBlank()) {
                Label subtitleLabel = new Label(subtitle);
                subtitleLabel.getStyleClass().add("form-subtitle");
                header.getChildren().add(subtitleLabel);
            }
            if (hasToolbar) {
                HBox toolbarBar = new HBox(8);
                toolbarBar.setPadding(new Insets(8, 0, 0, 0));
                toolbarBar.getStyleClass().add("action-bar");
                for (JsonNode btn : toolbar) {
                    toolbarBar.getChildren().add(buildButton(btn));
                }
                header.getChildren().add(toolbarBar);
            }
            root.setTop(header);
        }

        // Children
        VBox body = new VBox(12);
        body.setPadding(new Insets(8, 0, 8, 0));
        JsonNode children = component.path("children");
        if (children.isArray()) {
            ComponentRenderer childRenderer = new ComponentRenderer(ctx);
            for (JsonNode child : children) {
                body.getChildren().add(childRenderer.render(child, state, data));
            }
        }

        ScrollPane scroll = new ScrollPane(body);
        scroll.setFitToWidth(true);
        scroll.setStyle("-fx-background-color: transparent; -fx-background: transparent;");
        root.setCenter(scroll);

        // Bottom buttons
        JsonNode buttons = metadata.path("buttons");
        if (buttons.isArray() && !buttons.isEmpty()) {
            HBox bar = new HBox(8);
            bar.getStyleClass().add("action-bar");
            for (JsonNode btn : buttons) {
                bar.getChildren().add(buildButton(btn));
            }
            root.setBottom(bar);
        }

        return root;
    }

    private Button buildButton(JsonNode btn) {
        // ButtonDto uses "actionId"; fall back to "id" for ActionDto / other shapes
        String actionId = btn.path("actionId").asText(btn.path("id").asText(""));
        String label = btn.path("label").asText(actionId);
        Button button = new Button(label);
        boolean primary = "Primary".equalsIgnoreCase(btn.path("buttonStyle").asText(""));
        button.getStyleClass().add(primary ? "btn-primary" : "btn-default");
        button.setOnAction(e -> ctx.runAction(actionId, null));
        return button;
    }
}

File: frontend/app/intellij-plugin/src/main/kotlin/io/mateu/ijp/ui/PageRenderer.kt
package io.mateu.ijp.ui

import com.fasterxml.jackson.databind.JsonNode
import com.intellij.ui.components.JBLabel
import com.intellij.util.ui.JBUI
import io.mateu.ijp.api.arr
import io.mateu.ijp.api.text
import java.awt.BorderLayout
import java.awt.Color
import java.awt.Component
import java.awt.Font
import java.awt.FlowLayout
import javax.swing.JComponent
import javax.swing.JPanel

/**
 * Page — port of the JavaFX `PageRenderer`. Header (title/subtitle + toolbar buttons), optional
 * banners, the body `children`, and a bottom button bar.
 */
fun renderPage(r: ComponentRenderer, component: JsonNode, metadata: JsonNode, state: JsonNode, data: JsonNode): JComponent {
    val root = JPanel(BorderLayout(0, JBUI.scale(JBGap)))
    root.border = JBUI.Borders.empty(16)

    // Header (NORTH): title / subtitle / toolbar / banners.
    val header = verticalPanel(4)
    val title = metadata.text("title", metadata.text("pageTitle"))
    if (title.isNotBlank()) {
        val l = JBLabel(title)
        l.font = l.font.deriveFont(Font.BOLD, 20f)
        header.addStacked(l, 4)
    }
    val subtitle = metadata.text("subtitle")
    if (subtitle.isNotBlank()) {
        val l = JBLabel(subtitle)
        l.foreground = JBUI.CurrentTheme.Label.disabledForeground()
        header.addStacked(l, 8)
    }
    // Toolbar actions go to the native host toolbar (editor header / tool window title) when the
    // host provides one — the IntelliJ-idiomatic spot; otherwise render the inline button row.
    val toolbar = metadata.arr("toolbar")
    if (toolbar.isNotEmpty() && !r.ctx.publishToolbar(toolbar)) header.addStacked(buttonRow(r, toolbar), 8)
    for (banner in metadata.arr("banners")) header.addStacked(renderBanner(banner), 8)
    if (header.componentCount > 0) root.add(header, BorderLayout.NORTH)

    // Body (CENTER): a single child fills; several stack vertically.
    val children = if (component.path("children").isArray) component.path("children").toList() else emptyList()
    val body: JComponent = if (children.size == 1) {
        r.render(children[0], state, data)
    } else {
        val stack = verticalPanel()
        for (child in children) stack.addStacked(r.render(child, state, data), JBGap)
        stack
    }
    root.add(body, BorderLayout.CENTER)

    // Bottom button bar (SOUTH).
    val buttons = metadata.arr("buttons")
    if (buttons.isNotEmpty()) {
        val row = buttonRow(r, buttons)
        row.border = JBUI.Borders.emptyTop(JBGap)
        root.add(row, BorderLayout.SOUTH)
    }
    // The page's own CSS style caps the content width (e.g. `max-width:900px;margin:auto`, how the
    // web keeps forms readable) — anchored left, the IDE way, instead of centered.
    parseMaxWidth(component.text("style"))?.let { return MaxWidthPanel(it, root) }
    return root
}

internal fun buttonRow(r: ComponentRenderer, buttons: List<JsonNode>): JComponent {
    val row = JPanel(FlowLayout(FlowLayout.LEFT, JBUI.scale(JBGap), 0))
    row.isOpaque = false
    for (b in buttons) row.add(renderButton(r.ctx, b))
    return row
}

private fun renderBanner(banner: JsonNode): JComponent {
    val theme = banner.text("theme", "INFO").uppercase()
    val (bg, fg) = when (theme) {
        "SUCCESS" -> Color(0xE6, 0xF4, 0xEA) to Color(0x1E, 0x4A, 0x2B)
        "WARNING" -> Color(0xFD, 0xF6, 0xE3) to Color(0x6B, 0x53, 0x00)
        "DANGER", "ERROR" -> Color(0xFB, 0xE9, 0xE7) to Color(0x7A, 0x1E, 0x14)
        else -> Color(0xE8, 0xF0, 0xFE) to Color(0x0B, 0x3D, 0x91)
    }
    val panel = JPanel()
    panel.layout = javax.swing.BoxLayout(panel, javax.swing.BoxLayout.Y_AXIS)
    panel.background = bg
    panel.isOpaque = true
    panel.border = JBUI.Borders.empty(10, 12)
    val bTitle = banner.text("title")
    if (bTitle.isNotBlank()) {
        val l = JBLabel(bTitle)
        l.foreground = fg
        l.font = l.font.deriveFont(Font.BOLD)
        l.alignmentX = Component.LEFT_ALIGNMENT
        panel.add(l)
    }
    val desc = banner.text("description")
    if (desc.isNotBlank()) {
        val l = JBLabel(desc)
        l.foreground = fg
        l.alignmentX = Component.LEFT_ALIGNMENT
        panel.add(l)
    }
    return panel
}
